File: Leetcode/PY/Leetcode430-扁平化多级链表.py
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child

class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        t = head
        while t:
            ne = t.next
            if t.child:
                t.next = self.flatten(t.child)
                t.child = None
                t.next.prev = t
                while t.next:
                    t = t.next
            t.next = ne
            if ne:
                ne.prev = t
            t = ne
        return head

File: Leetcode/sbgo/Leetcode430-扁平化多级链表.go
package sbgo

func flatten(root *Node) *Node {
	t := root
	for t != nil {
		ne := t.Next
		if t.Child != nil {
			t.Next = flatten(t.Child)
			t.Next.Prev = t
			t.Child = nil
			for t.Next != nil {
				t = t.Next
			}
		}
		t.Next = ne
		if ne != nil {
			ne.Prev = t
		}
		t = ne
	}
	return root
}

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}
